flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -1016,6 +1034,47 @@ public CompletableFuture<CoordinationResponse> deliverCoordinationRequestToCoord
                                 operatorId, serializedRequest, timeout));
     }
 
+    @Override
+    public CompletableFuture<Void> reportJobClientHeartbeat(JobID jobId, Time timeout) {
+        if (!getJobManagerRunner(jobId).isPresent()) {
+            log.warn("Fail to find job {} for client.", jobId);
+        } else {
+            log.debug("Job {} receives client's heartbeat.", jobId);
+            markJobClientAliveness(jobId);
+            resetJobClientAlivenessCheck(jobId);
+        }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    private void markJobClientAliveness(JobID jobId) {
+        if (jobClientAlivenessFutures.containsKey(jobId)) {
+            jobClientAlivenessFutures.get(jobId).complete(null);
+        }
+    }
+
+    private void resetJobClientAlivenessCheck(JobID jobId) {
+        CompletableFuture<Void> clientAlivenessFuture = new CompletableFuture<>();
+        jobClientAlivenessFutures.put(jobId, clientAlivenessFuture);
+        FutureUtils.orTimeout(
+                clientAlivenessFuture,
+                jobClientHeartbeatTimeout,
+                TimeUnit.MILLISECONDS,
+                getMainThreadExecutor());
+        clientAlivenessFuture.whenComplete(
+                (t, throwable) -> {
+                    if (throwable != null) {
+                        if (throwable instanceof TimeoutException) {
+                            log.warn(
+                                    "Haven't receive aliveness from job client and cancel the job {}.",
+                                    jobId);
+                            cancelJob(jobId, webTimeout);
```

Comment:
The future based approach seems to be an overkill for this feature. For each job and on each heartbeat, it creates a new feature, which seems to be expensive.

I'd suggest to simply record the last heartbeat timestamp of each job, and periodically check whether there's any job that has timed out. This is similar to how we check and release idle TMs. It may not trigger a timeout precisely on the configured time, but would be much cheaper.

Reply:
@xintongsong Agree with you and fixed.